opm/simulators/wells/BlackoilWellModelGeneric.cpp

```diff
@@ -1188,8 +1190,51 @@ updateAndCommunicateGroupData(const int reportStepIdx,
     // before we copy to well_state_nupcol_.
     this->wellState().updateGlobalIsGrup(comm_);
 
-    if (iterationIdx < nupcol) {
+    if (iterationIdx <= nupcol) {
         this->updateNupcolWGState();
+    } else {
+        for (const auto& gr_name : schedule().groupNames(reportStepIdx)) {
+            const Phase all[] = { Phase::WATER, Phase::OIL, Phase::GAS };
+            for (Phase phase : all) {
+                if (this->groupState().has_injection_control(gr_name, phase)) {
+                    if (this->groupState().injection_control(gr_name, phase) == Group::InjectionCMode::VREP || 
+                        this->groupState().injection_control(gr_name, phase) == Group::InjectionCMode::REIN) {
+                        const bool is_vrep = this->groupState().injection_control(gr_name, phase) == Group::InjectionCMode::VREP;
+                        const Group& group = schedule().getGroup(gr_name, reportStepIdx);
+                        const int np = this->wellState().numPhases();
+                        Scalar gr_rate_nupcol = 0.0;
+                        for (int phaseIdx = 0; phaseIdx < np; ++phaseIdx) {
+                            gr_rate_nupcol += WellGroupHelpers<Scalar>::sumWellPhaseRates(is_vrep,
+                                                    group,
+                                                    schedule(),
+                                                    this->nupcolWellState(),
+                                                    reportStepIdx,
+                                                    phaseIdx,
+                                                    /*isInjector*/ false);
+                        }
+                        Scalar gr_rate = 0.0;
+                        for (int phaseIdx = 0; phaseIdx < np; ++phaseIdx) {
+                            gr_rate += WellGroupHelpers<Scalar>::sumWellPhaseRates(is_vrep,
+                                                    group,
+                                                    schedule(),
+                                                    this->wellState(),
+                                                    reportStepIdx,
+                                                    phaseIdx,
+                                                    /*isInjector*/ false);
+                        }
+                        Scalar rel_change = std::abs( (gr_rate_nupcol - gr_rate) / (0.5*gr_rate_nupcol + 0.5*gr_rate));
```

Comment:
Is there a possibility both group rates in denominator could be zero here? Or would that situation lead to failure at an earlier stage?

Reply:
It is possible so I will add a check. Maybe we should have a "large epsilon" here since relative changes does not make much sense for very small numbers.